File: android/app/src/main/java/com/webhunt/app/domain/taxonomy/QueryAdapter.kt
package com.webhunt.app.domain.taxonomy

import com.webhunt.app.data.model.OnlineJobLead

data class NormalizedGeoTarget(
    val countryName: String,
    val isoCode: String,
    val jobicyGeo: String? = null // "usa", "uk", "canada", "emea", "apac", "latam"
)

data class AdaptedProviderQuery(
    val rawQuery: String,
    val cleanQuery: String,
    val tokens: List<String>,
    val primaryTag: String,
    val geo: NormalizedGeoTarget? = null,
    val isWorldwide: Boolean = true
)

object QueryAdapter {

    private val STOPWORDS = setOf(
        "a", "an", "and", "or", "in", "on", "at", "to", "for", "of", "with", "from",
        "by", "as", "is", "are", "was", "were", "the", "this", "that", "it",
        "remote", "jobs", "job", "work", "seeking", "wanted", "hiring", "openings",
        "roles", "role", "position", "positions", "looking", "candidate"
    )

    // Strict country and region mappings (prevents dangerous substring matches like "us" in "Mauritius" or "Cyprus")
    private val GEO_MAPPING: Map<String, NormalizedGeoTarget> = mapOf(
        // Kenya & East Africa
        "kenya" to NormalizedGeoTarget("Kenya", "KE", "emea"),
        "ke" to NormalizedGeoTarget("Kenya", "KE", "emea"),
        "nairobi" to NormalizedGeoTarget("Kenya", "KE", "emea"),
        "mombasa" to NormalizedGeoTarget("Kenya", "KE", "emea"),
        "uganda" to NormalizedGeoTarget("Uganda", "UG", "emea"),
        "tanzania" to NormalizedGeoTarget("Tanzania", "TZ", "emea"),
        "rwanda" to NormalizedGeoTarget("Rwanda", "RW", "emea"),
        "nigeria" to NormalizedGeoTarget("Nigeria", "NG", "emea"),
        "ghana" to NormalizedGeoTarget("Ghana", "GH", "emea"),
        "south africa" to NormalizedGeoTarget("South Africa", "ZA", "emea"),
        "africa" to NormalizedGeoTarget("Africa", "AF", "emea"),

        // USA & North America
        "united states" to NormalizedGeoTarget("United States", "US", "usa"),
        "usa" to NormalizedGeoTarget("United States", "US", "usa"),
        "us" to NormalizedGeoTarget("United States", "US", "usa"),
        "america" to NormalizedGeoTarget("United States", "US", "usa"),
        "canada" to NormalizedGeoTarget("Canada", "CA", "canada"),
        "ca" to NormalizedGeoTarget("Canada", "CA", "canada"),

        // United Kingdom & Europe
        "united kingdom" to NormalizedGeoTarget("United Kingdom", "UK", "uk"),
        "uk" to NormalizedGeoTarget("United Kingdom", "UK", "uk"),
        "england" to NormalizedGeoTarget("United Kingdom", "UK", "uk"),
        "britain" to NormalizedGeoTarget("United Kingdom", "UK", "uk"),
        "germany" to NormalizedGeoTarget("Germany", "DE", "emea"),
        "france" to NormalizedGeoTarget("France", "FR", "emea"),
        "netherlands" to NormalizedGeoTarget("Netherlands", "NL", "emea"),
        "spain" to NormalizedGeoTarget("Spain", "ES", "emea"),
        "italy" to NormalizedGeoTarget("Italy", "IT", "emea"),
        "switzerland" to NormalizedGeoTarget("Switzerland", "CH", "emea"),
        "ireland" to NormalizedGeoTarget("Ireland", "IE", "emea"),
        "europe" to NormalizedGeoTarget("Europe", "EU", "emea"),
        "eu" to NormalizedGeoTarget("Europe", "EU", "emea"),
        "emea" to NormalizedGeoTarget("Europe, Middle East, Africa", "EMEA", "emea"),

        // Asia & Pacific
        "india" to NormalizedGeoTarget("India", "IN", "apac"),
        "singapore" to NormalizedGeoTarget("Singapore", "SG", "apac"),
        "japan" to NormalizedGeoTarget("Japan", "JP", "apac"),
        "australia" to NormalizedGeoTarget("Australia", "AU", "apac"),
        "new zealand" to NormalizedGeoTarget("New Zealand", "NZ", "apac"),
        "apac" to NormalizedGeoTarget("Asia-Pacific", "APAC", "apac"),
        "asia" to NormalizedGeoTarget("Asia", "ASIA", "apac"),

        // Latin America
        "brazil" to NormalizedGeoTarget("Brazil", "BR", "latam"),
        "mexico" to NormalizedGeoTarget("Mexico", "MX", "latam"),
        "argentina" to NormalizedGeoTarget("Argentina", "AR", "latam"),
        "colombia" to NormalizedGeoTarget("Colombia", "CO", "latam"),
        "latam" to NormalizedGeoTarget("Latin America", "LATAM", "latam")
    )

    private val TECH_PRIORITY = listOf(
        "react", "nextjs", "next.js", "python", "javascript", "typescript",
        "node", "nodejs", "vue", "angular", "rust", "golang", "java", "ruby",
        "devops", "aws", "docker", "kubernetes", "flutter", "swift", "kotlin",
        "frontend", "backend", "fullstack", "design", "figma", "ai", "writer",
        "marketing", "sales", "support", "developer", "engineer"
    )

    /**
     * Strict geographic target normalization avoiding unsafe substring matches.
     * Guaranteed to never match "Mauritius" or "Cyprus" as "US".
     */
    fun normalizeGeoTarget(rawCountry: String?): NormalizedGeoTarget? {
        if (rawCountry.isNullOrBlank()) return null
        val cleaned = rawCountry.trim().lowercase()
        if (cleaned in setOf("worldwide", "global", "anywhere", "all", "worldwide / global")) {
            return null
        }

        // Exact match
        GEO_MAPPING[cleaned]?.let { return it }

        // Tokenized word-boundary match
        val words = cleaned.split(Regex("[\\s,]+"))
        for (w in words) {
            GEO_MAPPING[w]?.let { return it }
        }

        return null
    }

    /**
     * Extracts clean distinct keyword tokens from user query.
     * Normalizes punctuation, slashes, and stopwords while preserving compound tech tokens.
     */
    fun extractQueryTokens(rawQuery: String): List<String> {
        if (rawQuery.isBlank()) return emptyList()

        val normalized = rawQuery
            .lowercase()
            .replace(Regex("[\\\\/|,+&():;?!'\"\\[\\]{}<>]"), " ")
            .replace(Regex("\\s+"), " ")
            .trim()

        val words = normalized.split(" ")
        val tokens = mutableListOf<String>()

        for (w in words) {
            val clean = w.trim()
            if (clean.length > 1 && clean !in STOPWORDS) {
                tokens.add(clean)
            }
        }

        val rawLower = rawQuery.lowercase()
        if ("next.js" in rawLower || "nextjs" in rawLower) {
            if ("next.js" !in tokens) tokens.add("next.js")
            if ("nextjs" !in tokens) tokens.add("nextjs")
        }
        if ("react.js" in rawLower || "reactjs" in rawLower) {
            if ("react" !in tokens) tokens.add("react")
        }
        if ("node.js" in rawLower || "nodejs" in rawLower) {
            if ("node.js" !in tokens) tokens.add("node.js")
            if ("nodejs" !in tokens) tokens.add("nodejs")
        }
        if ("vue.js" in rawLower || "vuejs" in rawLower) {
            if ("vue" !in tokens) tokens.add("vue")
        }

        return tokens.distinct()
    }

    /**
     * Selects single authoritative slug for tag-based APIs (Jobicy, RemoteOK).
     */
    fun selectPrimaryTag(tokens: List<String>, rawQuery: String): String {
        if (tokens.isEmpty()) return "developer"

        for (tp in TECH_PRIORITY) {
            if (tokens.contains(tp)) {
                return when (tp) {
                    "next.js" -> "nextjs"
                    "node.js" -> "nodejs"
                    else -> tp
                }
            }
        }

        return tokens.firstOrNull() ?: "developer"
    }

    /**
     * Adapts raw search query and optional geography into structured query envelope.
     */
    fun adaptQuery(rawQuery: String, country: String? = null): AdaptedProviderQuery {
        val tokens = extractQueryTokens(rawQuery)
        val primaryTag = selectPrimaryTag(tokens, rawQuery)
        val geo = normalizeGeoTarget(country)
        val cleanQuery = if (tokens.isNotEmpty()) tokens.joinToString(" ") else rawQuery.trim()

        return AdaptedProviderQuery(
            rawQuery = rawQuery.trim(),
            cleanQuery = cleanQuery,
            tokens = tokens,
            primaryTag = primaryTag,
            geo = geo,
            isWorldwide = geo == null
        )
    }

    /**
     * Deterministically tests if an opportunity matches query tokens.
     * Uses multi-factor weighted checks across title, company, skills, tags, opportunityType, and description snippet.
     */
    fun leadMatchesSearchTokens(lead: OnlineJobLead, tokens: List<String>): Boolean {
        if (tokens.isEmpty()) return true

        val title = lead.title.lowercase()
        val company = lead.company.lowercase()
        val desc = (lead.descriptionSnippet ?: "").lowercase()
        val tagsStr = lead.tags.joinToString(" ").lowercase()
        val skillsStr = lead.skills.joinToString(" ").lowercase()
        val oppType = (lead.opportunityType ?: "").lowercase()
        val extraText = "$company $tagsStr $skillsStr $oppType"

        // 1. Direct title or metadata token match
        for (tok in tokens) {
            val tokLower = tok.lowercase()
            if (title.contains(tokLower) || extraText.contains(tokLower)) {
                return true
            }
        }

        // 2. Token match in description snippet (requires token length >= 3)
        for (tok in tokens) {
            val tokLower = tok.lowercase()
            if (tokLower.length >= 3 && desc.contains(tokLower)) {
                return true
            }
        }

        return false
    }
}

File: lib/taxonomy/query-adapter.ts
import { OnlineSearchParams } from "@/lib/types";

export interface NormalizedGeoTarget {
  countryName: string;
  isoCode: string;
  jobicyGeo?: "usa" | "uk" | "canada" | "emea" | "apac" | "latam";
}

export interface AdaptedProviderQuery {
  rawQuery: string;
  cleanQuery: string;
  tokens: string[];
  primaryTag: string;
  geo?: NormalizedGeoTarget;
  isWorldwide: boolean;
}

// Stopwords to exclude when extracting technical keywords
const STOPWORDS = new Set([
  "a", "an", "and", "or", "in", "on", "at", "to", "for", "of", "with", "from",
  "by", "as", "is", "are", "was", "were", "the", "this", "that", "it",
  "remote", "jobs", "job", "work", "seeking", "wanted", "hiring", "openings",
  "roles", "role", "position", "positions", "looking", "candidate"
]);

// Strict country and region mappings (prevents dangerous substring matches like "us" in "Mauritius")
const GEO_MAPPING: Record<string, NormalizedGeoTarget> = {
  // Kenya & East Africa
  kenya: { countryName: "Kenya", isoCode: "KE", jobicyGeo: "emea" },
  ke: { countryName: "Kenya", isoCode: "KE", jobicyGeo: "emea" },
  nairobi: { countryName: "Kenya", isoCode: "KE", jobicyGeo: "emea" },
  mombasa: { countryName: "Kenya", isoCode: "KE", jobicyGeo: "emea" },
  uganda: { countryName: "Uganda", isoCode: "UG", jobicyGeo: "emea" },
  tanzania: { countryName: "Tanzania", isoCode: "TZ", jobicyGeo: "emea" },
  rwanda: { countryName: "Rwanda", isoCode: "RW", jobicyGeo: "emea" },
  nigeria: { countryName: "Nigeria", isoCode: "NG", jobicyGeo: "emea" },
  ghana: { countryName: "Ghana", isoCode: "GH", jobicyGeo: "emea" },
  "south africa": { countryName: "South Africa", isoCode: "ZA", jobicyGeo: "emea" },
  africa: { countryName: "Africa", isoCode: "AF", jobicyGeo: "emea" },

  // USA & North America
  "united states": { countryName: "United States", isoCode: "US", jobicyGeo: "usa" },
  usa: { countryName: "United States", isoCode: "US", jobicyGeo: "usa" },
  us: { countryName: "United States", isoCode: "US", jobicyGeo: "usa" },
  america: { countryName: "United States", isoCode: "US", jobicyGeo: "usa" },
  canada: { countryName: "Canada", isoCode: "CA", jobicyGeo: "canada" },
  ca: { countryName: "Canada", isoCode: "CA", jobicyGeo: "canada" },

  // United Kingdom & Europe
  "united kingdom": { countryName: "United Kingdom", isoCode: "UK", jobicyGeo: "uk" },
  uk: { countryName: "United Kingdom", isoCode: "UK", jobicyGeo: "uk" },
  england: { countryName: "United Kingdom", isoCode: "UK", jobicyGeo: "uk" },
  britain: { countryName: "United Kingdom", isoCode: "UK", jobicyGeo: "uk" },
  germany: { countryName: "Germany", isoCode: "DE", jobicyGeo: "emea" },
  france: { countryName: "France", isoCode: "FR", jobicyGeo: "emea" },
  netherlands: { countryName: "Netherlands", isoCode: "NL", jobicyGeo: "emea" },
  spain: { countryName: "Spain", isoCode: "ES", jobicyGeo: "emea" },
  italy: { countryName: "Italy", isoCode: "IT", jobicyGeo: "emea" },
  switzerland: { countryName: "Switzerland", isoCode: "CH", jobicyGeo: "emea" },
  ireland: { countryName: "Ireland", isoCode: "IE", jobicyGeo: "emea" },
  europe: { countryName: "Europe", isoCode: "EU", jobicyGeo: "emea" },
  eu: { countryName: "Europe", isoCode: "EU", jobicyGeo: "emea" },
  emea: { countryName: "Europe, Middle East, Africa", isoCode: "EMEA", jobicyGeo: "emea" },

  // Asia & Pacific
  india: { countryName: "India", isoCode: "IN", jobicyGeo: "apac" },
  singapore: { countryName: "Singapore", isoCode: "SG", jobicyGeo: "apac" },
  japan: { countryName: "Japan", isoCode: "JP", jobicyGeo: "apac" },
  australia: { countryName: "Australia", isoCode: "AU", jobicyGeo: "apac" },
  "new zealand": { countryName: "New Zealand", isoCode: "NZ", jobicyGeo: "apac" },
  apac: { countryName: "Asia-Pacific", isoCode: "APAC", jobicyGeo: "apac" },
  asia: { countryName: "Asia", isoCode: "ASIA", jobicyGeo: "apac" },

  // Latin America
  brazil: { countryName: "Brazil", isoCode: "BR", jobicyGeo: "latam" },
  mexico: { countryName: "Mexico", isoCode: "MX", jobicyGeo: "latam" },
  argentina: { countryName: "Argentina", isoCode: "AR", jobicyGeo: "latam" },
  colombia: { countryName: "Colombia", isoCode: "CO", jobicyGeo: "latam" },
  latam: { countryName: "Latin America", isoCode: "LATAM", jobicyGeo: "latam" },
};

/**
 * Normalizes an arbitrary country or location string into an authoritative geo target.
 */
export function normalizeGeoTarget(rawCountry?: string | null): NormalizedGeoTarget | undefined {
  if (!rawCountry) return undefined;
  const cleaned = rawCountry.trim().toLowerCase();
  if (cleaned === "worldwide" || cleaned === "global" || cleaned === "anywhere" || cleaned === "all") {
    return undefined;
  }

  // Exact match first
  if (GEO_MAPPING[cleaned]) {
    return GEO_MAPPING[cleaned];
  }

  // Check word tokens in location string
  const words = cleaned.split(/[\s,]+/);
  for (const w of words) {
    if (GEO_MAPPING[w]) {
      return GEO_MAPPING[w];
    }
  }

  return undefined;
}

/**
 * Extracts clean, distinct keyword tokens from user search query.
 * Strips punctuation, slashes, and common query noise.
 */
export function extractQueryTokens(rawQuery: string): string[] {
  if (!rawQuery) return [];
  const normalized = rawQuery
    .toLowerCase()
    .replace(/[\\/|,+&():;?!'"[\]{}<>]/g, " ")
    .replace(/\s+/g, " ")
    .trim();

  const words = normalized.split(" ");
  const tokens: string[] = [];

  for (const w of words) {
    const clean = w.trim();
    if (clean.length > 1 && !STOPWORDS.has(clean)) {
      tokens.push(clean);
    }
  }

  // Preserve compound tech terms
  if (rawQuery.toLowerCase().includes("next.js") || rawQuery.toLowerCase().includes("nextjs")) {
    if (!tokens.includes("next.js")) tokens.push("next.js");
    if (!tokens.includes("nextjs")) tokens.push("nextjs");
  }
  if (rawQuery.toLowerCase().includes("react.js") || rawQuery.toLowerCase().includes("reactjs")) {
    if (!tokens.includes("react")) tokens.push("react");
  }
  if (rawQuery.toLowerCase().includes("node.js") || rawQuery.toLowerCase().includes("nodejs")) {
    if (!tokens.includes("node.js")) tokens.push("node.js");
    if (!tokens.includes("nodejs")) tokens.push("nodejs");
  }
  if (rawQuery.toLowerCase().includes("vue.js") || rawQuery.toLowerCase().includes("vuejs")) {
    if (!tokens.includes("vue")) tokens.push("vue");
  }

  return Array.from(new Set(tokens));
}

/**
 * Selects the single best keyword slug for tag-based APIs (Jobicy, Remote OK).
 */
export function selectPrimaryTag(tokens: string[], rawQuery: string): string {
  if (tokens.length === 0) return "developer";

  // High priority specific technology terms
  const techPriority = [
    "react", "nextjs", "next.js", "python", "javascript", "typescript",
    "node", "nodejs", "vue", "angular", "rust", "golang", "java", "ruby",
    "devops", "aws", "docker", "kubernetes", "flutter", "swift", "kotlin",
    "frontend", "backend", "fullstack", "design", "figma", "ai", "writer",
    "marketing", "sales", "support", "developer", "engineer"
  ];

  for (const tp of techPriority) {
    if (tokens.includes(tp)) {
      // Normalize next.js -> nextjs for tag slugs
      return tp === "next.js" ? "nextjs" : tp === "node.js" ? "nodejs" : tp;
    }
  }

  // Fallback to first non-generic token or developer
  return tokens[0] || "developer";
}

/**
 * Adapts raw search params into provider-specific parameters.
 * Supports both adaptQueryForProvider(params) and adaptQueryForProvider(providerKey, params).
 */
export function adaptQueryForProvider(
  paramsOrProviderKey: OnlineSearchParams | string,
  maybeParams?: OnlineSearchParams
): AdaptedProviderQuery {
  const params: OnlineSearchParams =
    typeof paramsOrProviderKey === "object"
      ? paramsOrProviderKey
      : maybeParams || { mode: "online", query: "" };

  const rawQuery = (params.query || "").trim();
  const tokens = extractQueryTokens(rawQuery);
  const primaryTag = selectPrimaryTag(tokens, rawQuery);
  const geo = normalizeGeoTarget(params.country);
  const isWorldwide = !geo;
  const cleanQuery = tokens.length > 0 ? tokens.join(" ") : rawQuery;

  return {
    rawQuery,
    cleanQuery,
    tokens,
    primaryTag,
    geo,
    isWorldwide,
  };
}

/**
 * Deterministically checks if a lead matches search tokens.
 * Overloaded to support:
 *   Signature A: leadMatchesSearchTokens(lead, tokens)
 *   Signature B: leadMatchesSearchTokens(tokens, title, descriptionSnippet, extraKeywords)
 */
export function leadMatchesSearchTokens(
  arg1: { title?: string | null; descriptionSnippet?: string | null; tags?: string[] | null; company?: string | null } | string[],
  arg2?: string[] | string | null,
  arg3?: string | null,
  arg4?: (string | null | undefined)[]
): boolean {
  let tokens: string[] = [];
  let title = "";
  let desc = "";
  let extraText = "";

  if (Array.isArray(arg1)) {
    // Signature B: (tokens, title, descriptionSnippet, extraKeywords)
    tokens = arg1;
    title = (arg2 as string || "").toLowerCase();
    desc = (arg3 || "").toLowerCase();
    extraText = Array.isArray(arg4) ? arg4.filter(Boolean).join(" ").toLowerCase() : "";
  } else {
    // Signature A: (lead, tokens)
    tokens = Array.isArray(arg2) ? arg2 : [];
    title = (arg1?.title || "").toLowerCase();
    desc = (arg1?.descriptionSnippet || "").toLowerCase();
    const company = (arg1?.company || "").toLowerCase();
    const tagsStr = Array.isArray(arg1?.tags) ? arg1.tags.join(" ").toLowerCase() : "";
    extraText = `${company} ${tagsStr}`;
  }

  if (!tokens || tokens.length === 0) return true;

  // 1. Direct match: Any token appears in title or extra text (company / tags / category)
  for (const tok of tokens) {
    if (title.includes(tok) || extraText.includes(tok)) {
      return true;
    }
  }

  // 2. Token match in description snippet: Requires at least 1 token match
  for (const tok of tokens) {
    if (tok.length >= 3 && desc.includes(tok)) {
      return true;
    }
  }

  return false;
}
